### PR TITLE
Simplify `IntVarValue`/`FloatVarValue`

### PR DESCRIPTION
**Pull request by compiler-errors**
r? @ghost

### CODE REVIEW

File: compiler/rustc_infer/src/infer/mod.rs [hunk 1]
```diff
@@ -1607,15 +1611,15 @@ impl<'tcx> InferCtxt<'tcx> {
                 // If `inlined_probe_value` returns a value it's always a
                 // `ty::Int(_)` or `ty::UInt(_)`, which never matches a
                 // `ty::Infer(_)`.
-                self.inner.borrow_mut().int_unification_table().inlined_probe_value(v).is_some()
+                !self.inner.borrow_mut().int_unification_table().inlined_probe_value(v).is_unknown()
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/mod.rs:1614]:
```suggestion
                self.inner.borrow_mut().int_unification_table().inlined_probe_value(v).is_known()
```
double negation is confusing esp given the length of the negated condition

File: compiler/rustc_infer/src/infer/mod.rs [hunk 2]
```diff
@@ -1607,15 +1611,15 @@ impl<'tcx> InferCtxt<'tcx> {
...
             }
 
             TyOrConstInferVar::TyFloat(v) => {
                 // If `probe_value` returns a value it's always a
                 // `ty::Float(_)`, which never matches a `ty::Infer(_)`.
                 //
                 // Not `inlined_probe_value(v)` because this call site is colder.
-                self.inner.borrow_mut().float_unification_table().probe_value(v).is_some()
+                !self.inner.borrow_mut().float_unification_table().probe_value(v).is_unknown()
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/mod.rs:1622]:
```suggestion
                self.inner.borrow_mut().float_unification_table().probe_value(v).is_known()
```

File: compiler/rustc_infer/src/infer/mod.rs [hunk 3]
```diff
@@ -1836,21 +1840,20 @@ impl<'a, 'tcx> ShallowResolver<'a, 'tcx> {
...
-                .inner
-                .borrow_mut()
-                .int_unification_table()
-                .probe_value(v)
-                .map(|v| v.to_type(self.infcx.tcx)),
+            ty::IntVar(v) => {
+                match self.infcx.inner.borrow_mut().int_unification_table().probe_value(v) {
+                    ty::IntVarValue::Unknown => None,
+                    ty::IntVarValue::IntType(ty) => Some(Ty::new_int(self.infcx.tcx, ty)),
+                    ty::IntVarValue::UintType(ty) => Some(Ty::new_uint(self.infcx.tcx, ty)),
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/mod.rs:1847]:
```suggestion
                    ty::IntVarValue::IntType(ty) => Some(Ty::new_int(self.infcx.tcx, ty)),
                    ty::IntVarValue::UintType(ty) => Some(Ty::new_uint(self.infcx.tcx, ty)),
                    ty::IntVarValue::Unknown => None,
```
the nittiest nit

File: compiler/rustc_infer/src/infer/mod.rs [hunk 4]
```diff
@@ -1836,21 +1840,20 @@ impl<'a, 'tcx> ShallowResolver<'a, 'tcx> {
...
-                .infcx
-                .inner
-                .borrow_mut()
-                .float_unification_table()
-                .probe_value(v)
-                .map(|v| v.to_type(self.infcx.tcx)),
+            ty::FloatVar(v) => {
+                match self.infcx.inner.borrow_mut().float_unification_table().probe_value(v) {
+                    ty::FloatVarValue::Unknown => None,
+                    ty::FloatVarValue::Known(ty) => Some(Ty::new_float(self.infcx.tcx, ty)),
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/mod.rs:1854]:
```suggestion
                    ty::FloatVarValue::Known(ty) => Some(Ty::new_float(self.infcx.tcx, ty)),
                    ty::FloatVarValue::Unknown => None,
```
part 2

**Review by lcnr — COMMENTED** [review]
i think you also have to update nll relate, even if i shouldn't use int and float vars iirc

I remember trying the same thing at some point, don't remember why i didn't land it. r=me after perf

**Comment by lcnr** [comment]
i do remember why 😁 because it worsens perf

**Comment by compiler-errors** [comment]
Running a few more experiments...

@bors try @rust-timer queue

**Comment by compiler-errors** [comment]
Cool lol

**Pull request merged**